istioctl/cmd/istioctl/istioctl_test.go

```diff
@@ -0,0 +1,582 @@
+				Version:   model.VirtualService.Version,
+			},
+			Spec: &networking.VirtualService{
+				Hosts:    []string{"*"},
+				Gateways: []string{"bookinfo-gateway"},
+				Http: []*networking.HTTPRoute{
+					{
+						Match: []*networking.HTTPMatchRequest{
+							{
+								Uri: &networking.StringMatch{
+									&networking.StringMatch_Exact{"/productpage"},
+								},
+							},
+							{
+								Uri: &networking.StringMatch{
+									&networking.StringMatch_Exact{"/login"},
+								},
+							},
+							{
+								Uri: &networking.StringMatch{
+									&networking.StringMatch_Exact{"/logout"},
+								},
+							},
+							{
+								Uri: &networking.StringMatch{
+									&networking.StringMatch_Prefix{"/api/v1/products"},
+								},
+							},
+						},
+						Route: []*networking.DestinationWeight{
+							{
+								Destination: &networking.Destination{
+									Host: "productpage",
+									Port: &networking.PortSelector{
+										Port: &networking.PortSelector_Number{80},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+)
+
+func TestGet(t *testing.T) {
+	cases := []struct {
+		configs       []model.Config
+		args          []string
+		wantOutput    string
+		wantError     *regexp.Regexp
+		wantException bool
+	}{
+		{ // case 0
+			[]model.Config{},
+			[]string{"routerules"},
+			`No resources found.
+`,
+			regexp.MustCompile("^$"),
+			false,
+		},
+		{ // case 1
+			testRouteRules,
+			[]string{"routerules"},
+			`NAME      KIND                        NAMESPACE
+a         RouteRule.config.v1alpha2   default
+b         RouteRule.config.v1alpha2   default
+c         RouteRule.config.v1alpha2   istio-system
+d         RouteRule.config.v1alpha2   default
+`,
+			regexp.MustCompile("^$"),
+			false,
+		},
+		{ // case 2
+			testGateways,
+			[]string{"gateways"},
+			`NAME               KIND                          NAMESPACE
+bookinfo-gateway   Gateway.networking.v1alpha3   default
+`,
+			regexp.MustCompile("^$"),
+			false,
+		},
+		{ // case 3
+			testVirtualServices,
+			[]string{"virtualservices"},
+			`NAME       KIND                                 NAMESPACE
+bookinfo   VirtualService.networking.v1alpha3   default
+`,
+			regexp.MustCompile("^$"),
+			false,
+		},
+		{
+			[]model.Config{},
+			[]string{"invalid"},
+			"",
+			regexp.MustCompile("^Usage:.*"),
+			true, // "istioctl get invalid" should fail
+		},
+	}
+
+	for i, c := range cases {
+		// Override the client factory used by main.go
+		clientFactory = mockClientFactoryGenerator(c.configs)
+
+		// run getCmd capturing output
+		stdOutput, errOutput, fErr := captureOutput(
+			func() error {
+				err := rootCmd.PersistentPreRunE(getCmd, c.args)
+				if err != nil {
+					return multierror.Prefix(err, "Could not prerun GET command")
+				}
+				err = getCmd.RunE(getCmd, c.args)
+				if err != nil {
+					return multierror.Prefix(err, "Could not run GET command")
+				}
+				return nil
+			})
+
```

Comment:
If possible, it would be good to invoke the subcommand via `Execute` to verify we've setup PersistentPreRunE correct. e.g.

```golang
rootCmd.SetArgs(c.args)
rootCmd.Execute()
```